infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/external/IngressPathTransformInverter.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.server.external;
+
+import static org.eclipse.che.workspace.infrastructure.kubernetes.server.external.ExternalServerExposer.PATH_TRANSFORM_PATH_CATCH;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.commons.annotation.Nullable;
+
+public class IngressPathTransformInverter {
+  private final Pattern pathTransformInverse;
+
+  @Inject
+  public IngressPathTransformInverter(
+      @Nullable @Named("che.infra.kubernetes.ingress.path_transform") String pathTransformFmt) {
+    this.pathTransformInverse = extractPathFromFmt(pathTransformFmt);
+  }
+
+  private static Pattern extractPathFromFmt(String pathTransformFmt) {
+    int refIdx = pathTransformFmt.indexOf(PATH_TRANSFORM_PATH_CATCH);
+    String pathMatchingFragment = "(.*)";
+
+    String transformed;
+    if (refIdx < 0) {
+      transformed = Pattern.quote(pathTransformFmt);
+    } else {
+      if (refIdx == 0) {
+        if (pathTransformFmt.length() > 2) {
+          transformed = pathMatchingFragment + Pattern.quote(pathTransformFmt.substring(2));
+        } else {
+          transformed = pathMatchingFragment;
+        }
+      } else {
+        String prefix = Pattern.quote(pathTransformFmt.substring(0, refIdx));
+        String suffix =
+            refIdx < pathTransformFmt.length() - 2
+                ? Pattern.quote(pathTransformFmt.substring(refIdx + 2))
+                : "";
+
+        transformed = prefix + pathMatchingFragment + suffix;
+      }
+    }
```

Comment:
It still takes a significant amount of time to figure out what this code block does, but I won't block merging over it.

Reply:
I've rewritten it to use regexes even though the code is ~ an order of magnitude slower. It is not on the critical path of any sort so if it helps readability, so be it. I didn't do the suggested string format for concatenation though because that makes it another order of magnitude slower while not adding much to the readability.

I've ran some quick JMH tests and from this point on I will strongly object to using String.format for anything else but to compose error messages in catch blocks (which is what we use it for mostly, fortunately) ;)